### DISCUSSION TITLE
docs/design/project-structure

### DISCUSSION THREAD

**Discussion by giscus** [gogf/gf-site#24] [General]
# docs/design/project-structure

工程目录设计是代码分层设计的进一步落地，建议您先仔细阅读： 代码分层设计

https://goframe.org/docs/design/project-structure

**Comment by dfkgithub** [comment]
实际开发中常常纠结于controller与service的职责划分，也见识过很多别人设计的划分模式，比如所有逻辑都放service层controller仅作转发，或者在controller仅作转发，但service中再分出一个manager层。也有把绝大部分业务逻辑写在controller的，但可能一个controller方法数百行，一个controller类数千上万行。本章节提到的划分标准，是一个很有益的参考。需要特别强调的是，绝对不能在controller层跨层调用dao层，否则会留下难以处理的技术债。

**Reply by goodbugood** [reply]
controller 依赖了 http 协议的 api.Req，把业务逻辑写到 controller 里方向就是错的，到时候怎么复用。

**Comment by wangle201210** [comment]
文中对model层的使用前后描述不一致～

**Reply by tianyw0** [reply]
应该已经改了现在 两张截图说明了问题

**Comment by wangle201210** [comment]
gf gen dao 是不是可以加一个参数选项
开启后直接生成通用的增删改查及分页？

**Comment by yinglunfeng** [comment]
“model 中的数据结构不应该直接暴露给外部使用”，但是举的例子里面api数据结构又直接引用了/model/entity里面的实体模型，这个前后矛盾，该如何理解

**Reply by gqcn** [reply]
为了偷懒图方便呗。这里并没有做严格的限制，大多数场景其实都是直接引用`entity`的，其实并不是很好。但是管他喵那么多呢，谁能保证自己能在这个公司活得比项目更久？

**Reply by tianyw0** [reply]
最新的文档应该已经没有你所说的问题了，两张截图说明了问题。

**Reply by tianyw0** [reply]
这里我想讨论的问题是如何从规则上限制 API 层对model 的引用，目前看这个即使是 model 放在 internal 文件夹中也是没有用的，照样能被 API 引用。

**Comment by yinglunfeng** [comment]
亲，您的邮件我已收到，我看到后会尽快恢复您的！

**Comment by cui3093** [comment]
刚看到internal目录下已经没有`logic`目录了，但是通过`gf init`初始化项目，internal下依然存在`logic`，版本：2.9.7；
`二、请求分层流转`的图片中依然还有`logic`层，请问该怎么进行最佳实践呢？

**Comment by yinglunfeng** [comment]
亲，您的邮件我已收到，我看到后会尽快恢复您的！

**Comment by sitewang** [comment]
controller 是主要服务于API的业务；主要在比如需要查询订单及所属订单用户手机号时；就需要调用service下的订单模块和用户模块拿到数据；组合输出到API；这个DDD设计思想；

**Comment by yinglunfeng** [comment]
亲，您的邮件我已收到，我看到后会尽快恢复您的！

**Comment by goodbugood** [comment]
关于对 controller 层的描述：“controller 用于接收 api 的输入，可以直接在 controller 中实现业务逻辑，或者调用一个或多个 service 包实现业务逻辑，将执行结果封装为约定的 api 输出数据结构。”。我有个疑问，直接在 controller 中实现业务逻辑不太好吧。你的 controller 依赖了为 http 协议定义的 api 层参数。而你的项目后期可能还会有 grpc 协议，定时任务，消费者这些，如果要复用 controller 里你写的业务逻辑，岂不是都要将与 http 协议无关的参数转成 api.Req 才能复用？

**Comment by yinglunfeng** [comment]
亲，您的邮件我已收到，我看到后会尽快恢复您的！
